pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,721 @@
+}
+
+// SetDefaults configures the AttacherConfig with default values for those fields for which the compiler
+// defaults are not enough
+func (ac *AttacherConfig) SetDefaults() {
+	if ac.ScanTerminatedProcessesInterval == 0 {
+		ac.ScanTerminatedProcessesInterval = 30 * time.Minute
+	}
+
+	if ac.ProcRoot == "" {
+		ac.ProcRoot = "/proc"
+	}
+
+	if ac.EbpfConfig == nil {
+		ac.EbpfConfig = ebpf.NewConfig()
+	}
+}
+
+// ProbeManager is an interface that defines the methods that a Manager implements,
+// so that we can replace it in tests for a mock object
+type ProbeManager interface {
+	AddHook(string, *manager.Probe) error
+	DetachHook(manager.ProbeIdentificationPair) error
+	GetProbe(manager.ProbeIdentificationPair) (*manager.Probe, bool)
+}
+
+// FileRegistry is an interface that defines the methods that a FileRegistry implements, so that we can replace it in tests for a mock object
+type FileRegistry interface {
+	Register(namespacedPath string, pid uint32, activationCB, deactivationCB func(utils.FilePath) error) error
+	Unregister(uint32) error
+	Clear()
+	GetRegisteredProcesses() map[uint32]struct{}
+}
+
+// AttachCallback is a callback that is called whenever a probe is attached successfully
+type AttachCallback func(*manager.Probe, *utils.FilePath)
+
+// UprobeAttacher is a struct that handles the attachment of uprobes to processes and libraries
+type UprobeAttacher struct {
+	name         string
+	done         chan struct{}
+	wg           sync.WaitGroup
+	config       *AttacherConfig
+	fileRegistry FileRegistry
+	manager      ProbeManager
+	inspector    BinaryInspector
+
+	// pathToAttachedProbes maps a filesystem path to the probes attached to it. Used to detach them
+	// once the path is no longer used.
+	pathToAttachedProbes   map[string][]manager.ProbeIdentificationPair
+	onAttachCallback       AttachCallback
+	soWatcher              *sharedlibraries.EbpfProgram
+	handlesLibrariesCached *bool
+}
+
+// NewUprobeAttacher creates a new UprobeAttacher. Receives as arguments
+// the name of the attacher, the configuration, the probe manage (ebpf.Manager usually), a callback to be called
+// whenever a probe is attached (optional, can be nil), and the binary inspector to be used (e.g., to attach to
+// Go functions we need to inspect the binary in a different way)
+func NewUprobeAttacher(name string, config *AttacherConfig, mgr ProbeManager, onAttachCallback AttachCallback, inspector BinaryInspector) (*UprobeAttacher, error) {
+	config.SetDefaults()
+
+	ua := &UprobeAttacher{
+		name:                 name,
+		config:               config,
+		fileRegistry:         utils.NewFileRegistry(name),
+		manager:              mgr,
+		onAttachCallback:     onAttachCallback,
+		pathToAttachedProbes: make(map[string][]manager.ProbeIdentificationPair),
+		done:                 make(chan struct{}),
+		inspector:            inspector,
+	}
+
+	if ua.handlesLibraries() {
+		ua.soWatcher = sharedlibraries.NewEBPFProgram(config.EbpfConfig)
+	}
```

Comment:
should it be a singleton?

Reply:
If I understand the implementation of the channels in the perf handler correctly, it should. However, that'd force us also to make the uprobe attacher a singleton too, or to change the library watcher so that it can dispatch events to multiple receivers.